### PR TITLE
Firelocks and atmos optimizations

### PR DESCRIPTION
**Pull request by Zumorica**

### CODE REVIEW

**Comment by Zumorica** [comment]
Ready for review

File: Content.Server/Atmos/TileAtmosphere.cs [hunk 1]
```diff
@@ -1088,6 +1094,25 @@ private void HandleDecompressionFloorRip(float sum)
         private void ConsiderFirelocks(TileAtmosphere other)
         {
             // TODO ATMOS firelocks!
+            var reconsiderAdjacent = false;
```
Comment (DrSmugleaf) [inline Content.Server/Atmos/TileAtmosphere.cs:1097]:
```suggestion
            var reconsiderAdjacent = false;
```

File: Content.Server/GameObjects/Components/Atmos/AirtightComponent.cs [hunk 2]
```diff
@@ -51,22 +70,51 @@ public override void ExposeData(ObjectSerializer serializer)
...
+        private void RotateEvent(RotateEvent ev)
+        {
+            if (!_rotateAirBlocked || ev.Sender != Owner || ev.NewRotation == ev.OldRotation || AirBlockedDirection == AtmosDirection.Invalid)
+                return;
+
+            var diff = ev.NewRotation - ev.OldRotation;
+
+            var newAirBlockedDirs = AtmosDirection.Invalid;
+
+            // When we make multiZ atmos, special case this.
```
Comment (DrSmugleaf) [inline Content.Server/GameObjects/Components/Atmos/AirtightComponent.cs:105]: For ctrl+f purposes

```suggestion
            // TODO ATMOS MULTIZ When we make multiZ atmos, special case this.
```

File: Content.Server/GameObjects/Components/Atmos/AirtightComponent.cs [hunk 3]
```diff
@@ -111,13 +156,18 @@ private void OnTransformMove()
...
-            var mapIndices = Owner.Transform.Coordinates.ToMapIndices(_entityManager, _mapManager);
-            UpdatePosition(Owner.Transform.GridID, mapIndices);
+            if (Owner.TryGetComponent(out SnapGridComponent? snapGrid))
+                UpdatePosition(Owner.Transform.GridID, snapGrid.Position);
         }
 
         private void UpdatePosition(GridId gridId, MapIndices pos)
         {
-            EntitySystem.Get<AtmosphereSystem>().GetGridAtmosphere(gridId)?.Invalidate(pos);
+            var gridAtmos = EntitySystem.Get<AtmosphereSystem>().GetGridAtmosphere(gridId);
```
Comment (DrSmugleaf) [inline Content.Server/GameObjects/Components/Atmos/AirtightComponent.cs:165]: The entity system here could be cached if you are concerned about the amount of times this method is going to be called.

File: Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+            if (tool.HasQuality(ToolQuality.Prying))
+            {
+                var holdingPressure = IsHoldingPressure();
+                var holdingFire = IsHoldingFire();
+
+                if (State == DoorState.Closed)
+                {
+                    if(holdingPressure)
+                        _notifyManager.PopupMessage(Owner, eventArgs.User, "A gush of air blows in your face... Maybe you should reconsider.");
```
Comment (DrSmugleaf) [inline Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs:90]: Replace this with `Owner.PopupMessage(eventArgs.User, "...");` and remove the dependency

File: Content.Server/Utility/GridTileLookupHelpers.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+{
+    public static class GridTileLookupHelpers
+    {
+        /// <summary>
+        ///     Helper that returns all entities in a turf very fast.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static IEnumerable<IEntity> GetEntitiesInTileFast(this TileRef turf)
+        {
+            var gridTileLookup = EntitySystem.Get<GridTileLookupSystem>();
```
Comment (DrSmugleaf) [inline Content.Server/Utility/GridTileLookupHelpers.cs:20]: Given that this is "Fast" and you are even inlining it you might as well allow this system to be passed as a nullable parameter.

File: Content.Shared/Maps/TurfHelpers.cs [hunk 6]
```diff
@@ -120,13 +123,40 @@ public static bool TryGetTileRef(this EntityCoordinates coordinates, [NotNullWhe
         /// <summary>
         ///     Helper that returns all entities in a turf.
         /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static IEnumerable<IEntity> GetEntitiesInTile(this TileRef turf, bool approximate = false)
         {
             var entityManager = IoCManager.Resolve<IEntityManager>();
```
Comment (DrSmugleaf) [inline Content.Shared/Maps/TurfHelpers.cs:129]: This can be made a nullable parameter as well.

**Review by DrSmugleaf — CHANGES_REQUESTED** [review]

File: Content.Server/GameObjects/Components/Atmos/AirtightComponent.cs (hunk 3 above)
Comment (Zumorica) [reply Content.Server/GameObjects/Components/Atmos/AirtightComponent.cs]: done

File: Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs (hunk 4 above)
Comment (Zumorica) [reply Content.Server/GameObjects/Components/Atmos/FirelockComponent.cs]: done

File: Content.Server/Utility/GridTileLookupHelpers.cs (hunk 5 above)
Comment (Zumorica) [reply Content.Server/Utility/GridTileLookupHelpers.cs]: true, done

File: Content.Shared/Maps/TurfHelpers.cs (hunk 6 above)
Comment (Zumorica) [reply Content.Shared/Maps/TurfHelpers.cs]: done

**Pull request merged**